### PR TITLE
Make Tracks identity cookie cache compatible

### LINKED ISSUE #8162: Make Tracks cookie cache compatible
### Describe the bug

`tk_ai` is set by WooPay Tracks integration on page load [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-checkout.php#L203) (#6904). Typically, the cookie is set via an Ajax requests or as a result of a hook. In this case, the cookie is set, and its value is immediately printed onto the page to make it available if/when a user navigates to WooPay. 
As pointed out in p6q8Tx-4j4-p2#comment-8955, setting the cookie in PHP during a page load bypasses the page cache in certain environments. One proposed solution is to set the cookie in frontend. We should review the cookie usage to make sure whether it needs to be set in the initial page load or whether it can be moved elsewhere.

### Additional context
p2: p6q8Tx-4j4-p2#comment-8955
Slack: p1707306910715269-slack-CGGCLBN58

### PR DESCRIPTION
**Pull request by malithsen**
Fixes #8162 

#### Changes proposed in this Pull Request

This PR addresses an issue with the `tk_ai` cookie and its impact on page caching. Previously, the `tk_ai` cookie was set on page load, which made pages uncacheable. This change defers the setting of this cookie until it's actually needed, thus improving cacheability. The solution involves fetching the Tracks identity by checking for the presence of the tk_ai cookie and setting it via an Ajax request if it's not present.

See p6q8Tx-4j4-p2#comment-8955 for a detailed explanation of the issue. 

**Key changes:**

- Removal of the immediate setting of the `tk_ai` cookie on page load.
- Introduction of `getTracksIdentity` function in `client/tracks/index.ts` to asynchronously fetch the Tracks identity.
- Modification of iframe sources in `woopay-connect-iframe.js` and `woopay-express-checkout-button.js` to include the `tracksUserIdentity` parameter only after the Tracks identity is obtained.
- Removal of duplicated attempts to set `tk_ai` cookie in `class-woopay-tracker.php`

#### Testing instructions

**Setup**
- Enable WooPay.
- Make sure store country is set to US in WooCommerce -> Settings -> General.

**Reproduce the issue in `develop`**
1. As a shopper, visit the site in an incognito winow.
2. Open dev tools and make sure the `tk_ai` cookie is not set.
3. Add a product to the cart and visit the cart page.
4. Notice that the `Set-Cookie` header in response headers of the initial page request.

SCR-20240214-kujk

**Test the fix in this branch**
1. Redo the above steps and make sure the Set-Cookie header is absent in the initial page load request. `tk_ai` cookie will still be set via an Ajax request.
2. Remove the `tk_ai` cookie.
3. Visit a product page and make sure `Set-Cookie` header is absent on the response headers.
4. Remove the cookie once more.
5. Visit the checkout page and make sure `Set-Cookie` header is absent on the response headers.

**Regression tests**
* Test whether the Tracks User ID is transferred to WooPay in the first-party auth flow.
1. Add a breakpoint [here](https://github.com/Automattic/woopay/blob/be1a12b871b431d11db0d88d731d64a68abf50d2/src/Connect/ConnectPage.php#L85).
2. As a shopper, add a product to the cart and visit the cart page.
3. Above breakpoint should be reached. 
4. Make sure the `tracks_user_identity` field inside the `env_data` variable contains `_ui` and `_ut` entries.

* Test whether Tracks User ID is transferred to WooPay in the iframe flow
1. As a shopper, add a product to the cart and go to the checkout page.
2. Enter a WooPay registered email address.
3. Wait for the OTP iframe to show up.
4. Inspect the iframe and make sure that the `tracksUserIdentity` is passed on to the iframe as a URL parameter.

SCR-20240214-ljwo

* Test the Direct Checkout Flow
1. Make sure Direct Checkout feature is enabled for your store.
2. As a shopper add a product to the cart and visit the cart page.
3. Open dev tools and ensure the Direct Checkout iframe contains the `tracksUserIdentity`

SCR-20240214-llib

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/class-wc-payments-checkout.php [hunk 1]
```diff
@@ -230,10 +230,6 @@ public function get_payment_fields_js_config() {
 		}
 		$payment_fields['enabledBillingFields'] = $enabled_billing_fields;
 
-		if ( WC_Payments::woopay_tracker()->should_enable_tracking() ) {
-			$payment_fields['tracksUserIdentity'] = WC_Payments::woopay_tracker()->tracks_get_identity( get_current_user_id() );
-		}
-
```
Comment (malithsen) [inline includes/class-wc-payments-checkout.php:238]: This was the main culprit. This code is run during page load. `tracks_get_identity` will set a `tk_ai` cookie if it's not already present.

File: includes/class-woopay-tracker.php [hunk 2]
```diff
@@ -368,10 +380,6 @@ public function tracks_get_identity( $user_id ) {
 			add_user_meta( $user_id, 'jetpack_tracks_anon_id', $anon_id, false );
 		}
 
-		if ( ! isset( $_COOKIE['tk_ai'] ) && ! headers_sent() ) {
-			setcookie( 'tk_ai', $anon_id );
-		}
-
```
Comment (malithsen) [inline includes/class-woopay-tracker.php:374]: This cookie is already set inside the `\Jetpack_Tracks_Client::get_anon_id()` call, so this is redundant.

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8185` or branch name `fix/tk-ai-caching` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/tk-ai-caching) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 1fee8078724f93d9a83d74dd1a2fbb650f0049e4
- Build time: 2024-02-15 04:57:32 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by alefesouza — APPROVED** [review]
Thanks for working on this @malithsen! Code looks good and works as expected.

**Pull request merged**